Sign in with Apple with SwiftUI (#133)
xmartlabs/blog: _posts/2020-08-08-sign-in-with-apple-with-swiftui.md

### How do we get this data?

App invokes `onCompletion` closure with the result of the authorization request. If successful we receive the authorization credential mentioned before.

```swift
onCompletion: { (result) in    
	switch result {
	case .success(let authorization):
	    //Handle autorization
	    if let appleIDCredential = authorization.credential as? ASAuthorizationAppleIDCredential {
	        let userId = appleIDCredential.user
	        let identityToken = appleIDCredential.identityToken
	        let authCode = appleIDCredential.authorizationCode
	        let email = appleIDCredential.email
	        let givenName = appleIDCredential.fullName?.givenName
	        let familyName = appleIDCredential.fullName?.familyName
	        let state = appleIDCredential.state
	        // Here you have to send the data to the backend and according to the response let the user get into the app.
	    }
	    break
	case .failure(let error):
	    //Handle error
	    break
	}
}
```

It’s worth highlighting that we only receive all the information if it’s a new user, otherwise we don’t receive `fullName` neither `email`. Email data is always included in the identityToken JWT data.

> If something goes wrong the error is shown to the user by Apple SDK through a dialog. Anyway, we might still be interested in handling the error, for instance, in order to log it.

Reviewer: ```suggestion
If something goes wrong the error is shown to the user by Apple SDK through a dialogue. Nevertheless, we might still be interested in manually handling the error (in order to log in, for instance).
```

### Handling changes

Users could revoke permission for your app from *Setting → Apple ID → Password & Security → Apps Using Your Apple ID*.

<div style="text-align: center"><img width="60%" src="/images/sign-in-with-apple-swiftui/revoke.png" /></div>

Apple provides a way to know when an authentication change happens through `ASAuthorizationAppleIDProvider.credentialRevokedNotification`  notification.

> To handle the change we need to register for the notification:

Reviewer: ```suggestion
To cope with the change we need to register for the notification:
```

```swift
NotificationCenter.default.addObserver(self, selector: #selector(appleIDCredentialRevoked(_:)), name: ASAuthorizationAppleIDProvider.credentialRevokedNotification, object: nil)
```

> We can check the credential state with `getCredentialState` as the notification handler in the code below shows. Remember that we should have saved the user identifier in our app keychain as we told earlier.

Reviewer: ```suggestion
We can check the credential state with `getCredentialState`, as the notification handler in the code below shows. Remember that we should have saved the user identifier in our app keychain as we discussed earlier.
```

```swift
// ASAuthorizationAppleIDProvider.credentialRevokedNotification handler
@objc func appleIDCredentialRevoked(_ notification: Notification) {
    let appleIDProvider = ASAuthorizationAppleIDProvider()
    appleIDProvider.getCredentialState(forUserID: userIdentifier) { (credentialState, error) in
        switch credentialState {
        case .authorized:
            // The Apple ID credential is valid.
            break
        case .revoked:
            // The Apple ID credential is revoked. Sign out.
            break
        case .notFound:
            // No credential was found. Show login UI.
            break
        case .transferred:
            // The application was transferred from one development team to another.

> 					  // You can use the same code used to authenticate the user adding the locally saved user identifier in the request.

Reviewer: Why is this indented so far ahead?

        default:
            break
        }
    }
}
```

> **Transferred** is the less common credential state since it represents that the application was recently transferred from one development team to another. **UserIDs** are unique within a development team so it must be migrated. For more information about how to migrate, see Apple official documentation.

Reviewer: Link to that section of the documentation

If the user previously decided to use a private email relay for their account you can also receive these notifications:

- **email-disabled:** when a user has decided to stop receiving emails on their email relay.
- **email-enabled:** when the user opted back into receiving emails.

Here we can see how this JWT looks like:

<div style="text-align: center"><img width="60%" src="/images/sign-in-with-apple-swiftui/JWT-server-to-server-notification.png" /></div>

## Register your email domains

As we mentioned before, to communicate with app users who tap the *hide my email* option we must register the app email server domain. You must have already configured Sender Policy Framework (SPF) in order to use it at this point.

In order to configure your email domains enter to **Apple Developer Program**. Go to *Certificates, Identifiers & Profile → More* and tap *Configure* button

<img width="100%" src="/images/sign-in-with-apple-swiftui/emailComunication.png" />

Tap *+* to register your email sources. In *Domains and Subdomains* section add your domain name, click *Next* and *Register*

<img width="100%" src="/images/sign-in-with-apple-swiftui/registerEmailSources.png" />

> > The register will fail if you don't use SPF. If you're using Google to send mails, [here](https://support.google.com/a/answer/33786?hl=en) you have a configuration guide.

Reviewer: ```suggestion
> Registering the domain will fail if you don't use SPF. If you're using Google to send mails, [here](https://support.google.com/a/answer/33786?hl=en) you have a configuration guide.
```

After you get your domain registered, click *Download* and place the file in the specified location (https://YourDomain/.well-known/apple-developer-domain-association.txt) and click *Verify*.

<img width="100%" src="/images/sign-in-with-apple-swiftui/downloadRegister.jpeg" />

You will see a green checkmark if your domain passes the verification:

<img width="100%" src="/images/sign-in-with-apple-swiftui/checkDomain.jpeg" />

## Upgrading users to Sign in with Apple

> *Sign in with Apple* provides multiple benefits to current app users as we have discussed, so at a certain point some app users would be interested in start using Sign in with Apple over its current authentication system. Before iOS 14, there wasn't a simple way to do so, normally we ended up losing their account information or having a duplicated user.

Reviewer: ```suggestion
*Sign in with Apple* provides multiple benefits to current app users as we have discussed, so, at a certain point, some app users would be interested in starting using Sign in with Apple over its current authentication system. Before iOS 14, there wasn't a simple way to do so, normally we ended up losing their account information or having a duplicated user.
```

> iOS 14 introduced a new API to help users migrate to Sign in with Apple. This new API provides you the support for adding your custom UI in case any custom security verification needs to be completed before asking for a Sign in with Apple credential.

Reviewer: ```suggestion
iOS 14 introduced a new API to help users migrate to Sign in with Apple. This new API provides you with the support for adding your custom UI in case any custom security verification needs to be completed before asking for a Sign in with Apple credential.
```

## Aspects to keep in mind

> There are some aspects you should consider if you’ll integrate it.

Reviewer: ```suggestion
There are some aspects you should consider if you decide to integrate Sign in with Apple.
```

> As we mentioned before, developers only receive email and full name once, so if there is a connection issue and you don’t save this data locally you won’t be able to recover it.

Reviewer: ```suggestion
As we mentioned before, developers only receive the user's email and full name once, so if there is a connection issue and you don’t save this data locally you won’t be able to recover it.
```

If users choose the *hide my email* option, it could be difficult to identify the user since the app only holds its apple identifier and its apple private email. So any communication should be done through the app.

> Well, I hope now you have a better understanding about *Sign in with Apple*, its integration cost and if it’s suitable for your app!

Reviewer: ```suggestion
Well, I hope now you have a better understanding of *Sign in with Apple*, what's its integration cost, and if it’s desirable for your app!
```

> ***Are you integrating Sign in with Apple in your app and have learned something not covered in this post? Let me know in the comments. I’d be interested to add it to this blogpost.***

Reviewer: ```suggestion
***Are you integrating Sign in with Apple in your app, and did you face something not covered in this post? Let me know in the comments. I’d be interested to add it to this blogpost.***
```

Reviewer 2: esto estaba mal, hasta me suena mejor lo original @renzodgc

Reviewer: Decís por la coma? el "and have learned" me suena raro a mi, aunque capaz está correcto.

Sinó podría ser:
***Are you integrating Sign in with Apple in your app and did you encounter something that wasn't covered in this post? Let me know in the comments. I’d be interested to add it to this blogpost.***
```
